src/Raven.Server/Documents/DocumentDatabase.cs

```diff
@@ -178,13 +183,14 @@ private void ThrowOperationCancelled()
                 throw new OperationCanceledException("The database " + _parent.Name + " is shutting down");
             }
 
-
             public void Dispose()
             {
-                Interlocked.Decrement(ref _parent._usages);
-                if (_parent._databaseShutdown.IsCancellationRequested)
-                    _parent._waitForUsagesOnDisposal.Set();
+                long currentUsagesCount = 0;
+                if (_skipUsagesCount == false)
+                    currentUsagesCount = Interlocked.Decrement(ref _parent._usages);
 
+                if (_parent._databaseShutdown.IsCancellationRequested && currentUsagesCount == 0)
+                    _parent._waitForUsagesOnDisposal.Set();
```

Comment:
_skipUsagesCount is true only for changes requests

if _usages = 0 means that we don't have any more requests